### ISSUE TITLE
Use relational database schema for metadata

### ISSUE THREAD

**Issue by ninok** [diffractor/diffractor#176]
I was using Digikam previously to manage my picture collection before stumbling upon Diffractor which is doing exactly what I need and it's amazingly fast.

To check for duplicates or already imported files, I wrote some scripts that utilize the Digikam SQLite database:
https://github.com/ninok/digikam-utils
and I was trying to port those scripts to utilize the Diffractor SQLite database.
Unfortunately is the metadata not stored in a relational format, but inside a binary blob in the properties `item_properties.properties` column.
This makes it impossible to use SQL statements to for example find files by file hash.

Was there any reason to store the metadata in that way. Would it be possible to change that?

**Comment by ZacWalk** [comment]
Yes metadata is stored in the blob. Diffractor uses the sqlite db more like a cache and has its own internal query engine. That has some pros and cons but makes it difficuly to write a script to interact with it. I think changing the DB format would inconvenience a lot of users so I have no plans to do that.

Might be posible for me to add a function to export a CSV file containing all the metadata for the collection. Would that work for you?

I see you make use of hashes. Diffractor does not generate hashes for all files. Only for for small files or on demand for some comparisons. This would mean your reports would probabably need to generate their own hashes.

**Comment by ninok** [comment]
> Diffractor uses the sqlite db more like a cache and has its own internal query engine.

Okay, that's interesting. What was the reason for that design? Was SQLite to slow?

> I see you make use of hashes. Diffractor does not generate hashes for all files.

If it is not generating hashes, how is the duplicates feature working?

> Might be posible for me to add a function to export a CSV file containing all the metadata for the collection. Would that work for you?

Yeah, that would work, but that would be a clumsy workaround.
It would be nicer to have this implemented as part of the duplicates detection which is already there. As far as I can see in my image collection it's even detecting duplicates if the file hash is different, so I assume you are computing something like p hash.
I will dig a bit through the code and see if I can implement my script in Diffractor directly.

**Comment by ZacWalk** [comment]
I have never used Digikam or looked at its code. Any similarities are coincidental. Diffractor evolved out of my own workflow.

I have found over time it is just simplest to define a duplicate as a file with the same name, size and created date or matching crc32. Diffractor will only generate a crc32 for an item if it matches the size of another item and it wants to determin if they are the same binary files. It does generate a crc32 for all small images during import as this does not impact performance. 

For a while I did experiment with perceptual hashes and other ideas but that is not something I use in my workflow and users found it confusing. If I edit an image and save it as a new name, I dont consider it a duplicate.

**Comment by ZacWalk** [comment]
Closed in the latest version (v1.26.3). Please re-open if you still see any issues. Thanks!
